onionshare/onionshare
# Improved Flatpak and Snapcraft packaing

This PR, along with the `onionshare-2.4.1` branch of the [Flatpak repository](https://github.com/flathub/org.onionshare.OnionShare/tree/onionshare-2.4.1), fixes #1459.

In both Flatpak and Snapcraft, `snowflake-client` and `meek-client` are built from source, and I made some modifications to the code so that it finds them correctly

This also increases the security of the Flatpak package by revoking the `--filesystem=home:ro` permission that gives OnionShare read-only permission to the user's home dir. Instead, OnionShare only has permission to read and write to its config folder, as well as `~/OnionShare`. In share and website modes, there are now two separate buttons, "Add Files" and "Add Folder", and these use Flatpak portals to basically give access to files that are browsed for using those file selection dialogs.

Unfortunately, Flatpak doesn't support drag-and-drop portals, so if you're running the Flatpak package, I've disabled drag-and-drop. If you try to drag files in, it shows you this popup telling you to use the buttons instead:

![Screenshot from 2021-11-15 18-48-58](https://user-images.githubusercontent.com/156128/141887144-6db8461f-bd8c-4678-a0a3-3b225247f6cb.png)

This also fixes #1435 by switching from stem to [cepa](https://github.com/onionshare/cepa) 1.8.3, which is nice because it simplifies the Flatpak packaging some.

I haven't tested it because I haven't reproduced it, but this _should_ fix #1466 as well. In some OSes it looks like OnionShare crashes just trying to display the pretty ASCII art, so this commit https://github.com/onionshare/onionshare/commit/267e9eb1e43bb188e8925caecb71914130fb266f just does exception handling on the OnionShare banner, and if it hits an exception it skips the ASCII art and displays a much simpler banner.

This also _should_ fix #1437, but it's hard to be completely certain before the new snap is pushed in the Snap Store. But I was able to build a local snap, install it locally like `snap install --dangerous onionshare_2.4.1_amd64.snap`, which installs it with the sandbox enabled, and I'm not reproducing the crash. (I possibly fixed it by adding the `gnome-3-34` extension to the snap.)

## To test Flatpak

Get the code:

```
git clone https://github.com/flathub/org.onionshare.OnionShare
cd org.onionshare.OnionShare
git checkout onionshare-2.4.1
```

Build and install the Flatpak package, which takes awhile. You need `flatpak` and `flatpak-builder` to be installed.

```
flatpak-builder build --force-clean --install-deps-from=flathub --install --user org.onionshare.OnionShare.yaml
```

Then test it:

```
flatpak run org.onionshare.OnionShare
```

## To test Snapcraft

Make sure you're in the onionshare repo directory, and you have `snap` installed and also `snapcraft` installed (`snap install snapcraft --classic`).

Build the OnionShare snap:

```
snapcraft
```

This takes awhile. When it's done, install it:

```
sudo snap install --dangerous ./onionshare_2.4.1_amd64.snap
```

And run it:

```
/snap/bin/onionshare -v
```
